### DISCUSSION TITLE
Possible to use `validate_arguments` on a function with an arg called `validate`?

### DISCUSSION THREAD

**Discussion by cqfd** [pydantic/pydantic#4120] [Question] [answered]
I'm possibly being dumb here but I can't seem to figure out how to do something like this:

```.py
from pydantic import validate_arguments

@validate_arguments
def example(x: int, validate: bool) -> str:
  return "something"
```

Doing so gives `NameError: Field name "validate" shadows a BaseModel attribute; use a different field name with "alias='validate'".`

The particular function I'd like to decorate unfortunately can't change the name of its `validate` argument (let me know if there's some Python trick I'm missing though). I'm currently working around the error above by basically just re-implementing my own version of `validate_arguments`, where I append an `_` to all signature arguments before using them to build a pydantic model; this lets me use whatever argument names I want (🤞), but, yeah, I feel like `@validate_arguments` should already support arbitrary names—is there some built-in way to do what I'm trying to do above? (I haven't figured out how to use an alias yet.)

Thanks!

**Comment by uriyyo** [comment, marked as answer]
Hi @cqfd,

You can create intermediate decorator that will fix all the names for you:
```py
import inspect
from functools import wraps

from pydantic import validate_arguments, Field, BaseModel
from pydantic.fields import Undefined


def safe_validate_arguments(f):
    names_to_fix = {n for n in BaseModel.__dict__ if not n.startswith('_')}

    @wraps(f)
    def wrapper(*args, **kwargs):
        kwargs = {n[:-1] if n.removesuffix("_") in names_to_fix else n: v for n, v in kwargs.items()}
        return f(*args, **kwargs)

    def _create_param(p: inspect.Parameter) -> inspect.Parameter:
        default = Undefined if p.default is inspect.Parameter.empty else p.default
        return p.replace(name=f"{p.name}_", default=Field(default, alias=p.name))

    sig = inspect.signature(f)
    sig = sig.replace(parameters=[_create_param(p) if n in names_to_fix else p for n, p in sig.parameters.items()])

    wrapper.__signature__ = sig
    wrapper.__annotations__ = {f"{n}_" if n in names_to_fix else n: v for n, v in f.__annotations__.items()}

    return validate_arguments(wrapper, config={"allow_population_by_field_name": True})
```

Usage will be the same as with `validate_arguments`:
```py
@safe_validate_arguments
def example(x: int, validate: bool) -> str:
    return "something"
```
```py
>>> example(1, True)
something
```

**Comment by cqfd** [comment]
Aha, neat! That's nicer than my version, thank you 🙏
